File: src/main/java/com/polygloat/controllers/ImportController.java
package com.polygloat.controllers;

import com.polygloat.constants.Message;
import com.polygloat.dtos.PathDTO;
import com.polygloat.dtos.request.ImportDto;
import com.polygloat.exceptions.NotFoundException;
import com.polygloat.model.Language;
import com.polygloat.model.Permission;
import com.polygloat.model.Repository;
import com.polygloat.model.Source;
import com.polygloat.service.*;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.validation.SmartValidator;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@CrossOrigin(origins = "*")
@RequestMapping("/api/repository/{repositoryId}/import")
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class ImportController implements IController {

    private final TranslationService translationService;
    private final SourceService sourceService;
    private final RepositoryService repositoryService;
    private final SecurityService securityService;
    private final LanguageService languageService;
    private final SmartValidator validator;

    @PostMapping(value = "")
    public void doImport(@PathVariable("repositoryId") Long repositoryId, @RequestBody ImportDto dto) {
        securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.MANAGE);
        Repository repository = repositoryService.findById(repositoryId).orElseThrow(() -> new NotFoundException(Message.REPOSITORY_NOT_FOUND));

        Language language = languageService.getOrCreate(repository, dto.getLanguageAbbreviation());

        for (Map.Entry<String, String> entry : dto.getData().entrySet()) {
            Source source = sourceService.getOrCreateSource(repository, PathDTO.fromFullPath(entry.getKey()));
            translationService.setTranslation(source, language.getAbbreviation(), entry.getValue());
        }
    }
}

File: src/main/kotlin/io/tolgee/controllers/ImportController.kt
package io.tolgee.controllers

import io.swagger.v3.oas.annotations.tags.Tag
import io.tolgee.constants.Message
import io.tolgee.dtos.ImportDto
import io.tolgee.exceptions.NotFoundException
import io.tolgee.model.Permission
import io.tolgee.service.ImportService
import io.tolgee.service.RepositoryService
import io.tolgee.service.SecurityService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody
import java.io.OutputStream
import javax.validation.Valid


@RestController
@CrossOrigin(origins = ["*"])
@RequestMapping("/api/repository/{repositoryId}/import")
@Tag(name = "Import")
class ImportController(
        private val repositoryService: RepositoryService,
        private val securityService: SecurityService,
        private val importService: ImportService
) : IController {

    @PostMapping(value = [""])
    fun doImport(@PathVariable("repositoryId")
                 repositoryId: Long,
                 @RequestBody @Valid
                 dto: ImportDto
    ): ResponseEntity<StreamingResponseBody> {
        securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.MANAGE)
        val repository = repositoryService.getById(repositoryId).orElseThrow {
            NotFoundException(Message.REPOSITORY_NOT_FOUND)
        }

        val stream = StreamingResponseBody { out: OutputStream ->
            importService.import(repository, dto, out)
        }

        return ResponseEntity(stream, HttpStatus.OK)
    }
}
